traceql: fix signed-int parsing to correctly handle math.MinInt64 (issue #4623)

Syedowais312 opened the pull request:
# 🔧 Fix: Proper Handling of Signed Integers in TraceQL Lexer  
Closes: #4623

## Summary

This PR fixes incorrect parsing of signed integers in the TraceQL lexer.

Previously, expressions such as:

    { -9223372036854775808 }
    { +10 }
    { -5s }

were incorrectly parsed or caused unexpected failures due to improper unary sign handling and integer overflow behavior.

## What This PR Does

### ✔ 1. Correct signed integer parsing
- Supports unary `+` and `-` in valid contexts.
- Handles the `MinInt64` edge case (-9223372036854775808) safely.
- Validates numeric ranges to prevent silent overflows.

### ✔ 2. Correct signed duration parsing
Signed durations are now parsed consistently:

    -5s
    +10ms

using the existing `tryScanDuration` logic.

### ✔ 3. Fixes unary parsing context rules
Unary signs are now only allowed:
- at expression start,
- after `{`,
- after comparison operators,
- after commas,
- after unary `!`.

(Prevents incorrect parsing inside function argument lists.)

## Tests

Added and updated tests in `parse_test.go`:
- `{ -9223372036854775808 }`
- `{ +9223372036854775807 }`
- `{ -5s }`
- `{ +10ms }`

All related TraceQL tests pass.

## Why This Fix?

The previous logic failed in edge cases and mixed incorrectly with duration parsing.  
This fix is:
- simple,
- safe,
- correct,
- aligned with TraceQL grammar expectations.

## Result

TraceQL signed integer parsing is now robust, predictable, and fully tested.



## 🙏 Notes

If additional signed numeric types (float, duration, hex, etc.) need first-class support, I’d be happy to extend this work.

joe-elliott commented:
So this is quite a lot of complex code for such an intense edge case. Tests should keep it together, but we still need to weigh the value of the feature against the complexity of the code.

What do you think about an alternative solution like adding constants like `minInt`, `maxInt`? Could we do this in a cleaner way?

Syedowais312 commented:
Thanks for the review @joe-elliott!

I’ve updated the PR with the requested changes. The constant handling for `minInt`
and `maxInt` is now implemented cleanly in the parser, directly inside the `integer`
rule as suggested.

All generated parser files were updated using `make gen-traceql-local` and I’ve added
the corresponding tests and CHANGELOG entry.

Local build & full TraceQL test suite also pass successfully.

Please let me know if you’d like any follow-up changes!

Syedowais312 commented:
Thanks for pointing this out 
I've reset `CHANGELOG.md`  to upstream, restored the removed entries, and added a single changing entry for this TraceQL change under `main / unreleased` .

Please let me know if the wording looks good.

Syedowais312 commented:
Thanks for the clarification 👍  
I’ve kept the new error strings, updated the affected tests, and regenerated the TraceQL files using the CI image.  
All TraceQL tests are passing now.
